p2p/src/peers.rs

```diff
@@ -105,139 +104,32 @@ impl Peers {
 		Ok(peers.contains_key(&addr))
 	}
 
-	/// Get vec of peers we are currently connected to.
-	pub fn connected_peers(&self) -> Vec<Arc<Peer>> {
+	/// Iterator over our current peers.
+	/// This allows us to hide try_read_for() behind a cleaner interface.
+	/// PeersIter lets us chain various adaptors for convenience.
+	pub fn peers_iter(&self) -> PeersIter<impl Iterator<Item = Arc<Peer>>> {
 		let peers = match self.peers.try_read_for(LOCK_TIMEOUT) {
-			Some(peers) => peers,
+			Some(peers) => peers.values().cloned().collect(),
```

Comment:
We are potentially cloning a lot of `Arc`s here while we might only be interested in a few of them. Would it be possible to refactor `PeersIter` to borrow them instead, and only `clone` them whenever we use the `IntoIterator ` trait?

Reply:
I don't think this is really possible because peers are behind a `RwLock`. The cloning effectively takes a snapshot of the current state of our list of peers at that point in time. Otherwise we need to keep a reference to the lock and things get messy.

Let me take another look at this though.

Reply:
I think our peer management needs a larger overhaul to get rid of this cloning. I'm not convinced having this outer `RwLock` wrapping the peers is the right approach long term. 
Lets revisit this and tackle as a separate PR.

Reply:
Ok, agreed :+1: